### PR TITLE
Fix Flirt CRC and add Ff and afsb commands to check function signatures

### PR DESCRIPTION
**Pull request by wargio**
# DO NOT SQUASH

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

This PR requires this bin PR which breaks all the other PR tests. 
- https://github.com/rizinorg/rizin-testbins/pull/92

There were some bugs in the flirt code that made the whole signature set wrong.
- The CRC calculation was wrong, now it is correct and following flirt standard
- The way the signatures were applied was wrong
- The way the CRC was checked was wrong
- Now pat and sig files matches the same functions correctly with the same number of matches

The PR also adds 2 new commands:
- `afsb` generates the rizin signature of the function at the current offset (including the search signature to use with `/x` command)
image
- `Ff` outputs only the FLIRT pattern of the function at the current offset. 
image

Now the output is correct and matches 100% the output of IDA (with the exception of the byte mask which slightly differ from theirs)

This has been identified and tested with @ret2libc

### CODE REVIEW

File: librz/core/cconfig.c [hunk 1]
```diff
@@ -3676,8 +3676,9 @@ RZ_API int rz_core_config_init(RzCore *core) {
 		"aixar", "aout", "ar", "bin", "coff", "dos:com", "dos:com:old", "dos:exe", "dos:exe:old",
 		"dosdrv", "elf", "intelhex", "le", "loader", "lx", "moshex", "ne", "nlm", "omf", "omflib",
 		"pe", "pilot", "srec", "w32run", "zip", "all", "none", NULL);
-	SETB("flirt.sig.deflate", false, "enables/disables FLIRT zlib compression when creating a signature file (available only for .sig files)");
+	SETB("flirt.sig.deflate", false, "Enables/disables FLIRT zlib compression when creating a signature file (available only for .sig files)");
 	SETI("flirt.node.optimize", RZ_FLIRT_NODE_OPTIMIZE_MAX, "FLIRT optimization option when creating a signature file (none: 0, normal: 1, smallest: 2)");
+	SETB("flirt.ignore.unknown", true, "When enabled, on flirt creation it will ignore any function starting with `fcn.`");
```
Comment (XVilka) [inline librz/core/cconfig.c]: `flirt` -> `FLIRT`?

File: librz/core/cmd/cmd_analysis.c [hunk 2]
```diff
@@ -2189,6 +2189,106 @@ RZ_IPI RzCmdStatus rz_analysis_function_setbits_handler(RzCore *core, int argc,
...
+	return true;
+
+fail:
+	free(data);
+	return false;
+}
+
+RZ_IPI RzCmdStatus rz_analysis_function_signature_bytes_handler(RzCore *core, int argc, const char **argv, RzCmdStateOutput *state) {
+	RzAnalysisFunction *f = analysis_get_function_in(core->analysis, core->offset);
+	if (!f) {
```
Comment (XVilka) [inline librz/core/cmd/cmd_analysis.c:2227]:
```c
RZ_LOG_ERROR("core: Cannot find function at 0x%08" PFMT64x "\n", core->offset);
```
?

File: librz/sign/flirt.c [hunk 3]
```diff
@@ -201,7 +201,9 @@ typedef struct parse_status_t {
 // This is from flair tools flair/crc16.cpp
 // CRC-HDLC & CRC-16/X-25 produces the same but in LE format.
 #define POLY 0x8408
-ut16 flirt_crc16(const ut8 *data_p, size_t length) {
+RZ_API ut16 rz_flirt_crc16(RZ_NONNULL const ut8 *data_p, size_t length) {
```
Comment (XVilka) [inline librz/sign/flirt.c]: Please add Doxygen.

**Review by XVilka — APPROVED** [review]
LGTM with a few minor comments

File: librz/core/cmd/cmd_analysis.c [hunk 4]
```diff
@@ -2177,6 +2177,7 @@ RZ_IPI RzCmdStatus rz_analysis_function_setbits_handler(RzCore *core, int argc,
 	int bits = atoi(argv[1]);
 	RzAnalysisFunction *fcn = analysis_get_function_in(core->analysis, core->offset);
 	if (!fcn) {
+		RZ_LOG_ERROR("core: Cannot find function at 0x%08" PFMT64x "\n", core->offset);
```
Comment (p371jah) [inline librz/core/cmd/cmd_analysis.c]: `analysis_get_function_in` does it for you https://github.com/rizinorg/rizin/blob/1e5ae82bd38fe30ee041751e6a94d8dcf37004d5/librz/core/cmd/cmd_analysis.c#L182

File: test/db/cmd/cmd_signature [hunk 5]
```diff
@@ -202,11 +202,6 @@ EXPECT=<<EOF
 0x0044a030 142 flirt.sbrk
 0x0044a0c0 64 flirt.getpagesize
 0x0044a100 80 flirt.getdtablesize
-0x0044a2c0 327 flirt.trecurse
```
Comment (ret2libc) [inline test/db/cmd/cmd_signature:205]: did you check why these are removed?

File: test/db/cmd/cmd_signature [hunk 6]
```diff
@@ -397,7 +383,7 @@ aaa
 Fs bins/flirt/elf-x86/libcurl.a.sig
 EOF
 EXPECT=<<EOF
-Found 205 FLIRT signatures via bins/flirt/elf-x86/libcurl.a.sig
```
Comment (ret2libc) [inline test/db/cmd/cmd_signature:400]: This?

File: test/db/cmd/cmd_signature [hunk 7]
```diff
@@ -417,49 +403,51 @@ CMDS=<<EOF
 e flirt.sigdb.load.system=false
 e flirt.sigdb.load.home=false
 aaa
-Fs bins/flirt/elf-x86/test.symbols.pat~silence
-pdf @ main
+Fs bins/flirt/elf-x86/test.stripped.pat~silence
+af @ data.004011a5
+pdf @ data.004011a5
```
Comment (ret2libc) [inline test/db/cmd/cmd_signature:408]: why did you change the function being checked? Can you use that as well in the tests?

File: librz/core/cmd/cmd_flirt.c [hunk 8]
```diff
@@ -41,3 +41,73 @@ RZ_IPI RzCmdStatus rz_flirt_scan_handler(RzCore *core, int argc, const char **ar
...
+		RZ_LOG_ERROR("core: cannot allocate function buffer.\n");
+		return RZ_CMD_STATUS_ERROR;
+	}
+
+	if (!rz_io_read_at(core->io, func->addr, pattern, (int)size)) {
+		RZ_LOG_ERROR("core: couldn't read function %s at 0x%" PFMT64x ".\n", func->name, func->addr);
+		free(pattern);
+		return RZ_CMD_STATUS_ERROR;
+	}
+
```
Comment (ret2libc) [inline librz/core/cmd/cmd_flirt.c]: Does this assume a linear function? I think in cmd_analysis.c you iterated over the bbs instead of just getting the function linear size.

File: librz/core/cmd/cmd_flirt.c [hunk 9]
```diff
@@ -41,3 +41,73 @@ RZ_IPI RzCmdStatus rz_flirt_scan_handler(RzCore *core, int argc, const char **ar
...
+	ut16 crc16 = 0;
+	ut32 crclen = 0;
+	if (size > RZ_FLIRT_MAX_PRELUDE_SIZE) {
+		// the crc should be generated only for when the buffer is > RZ_FLIRT_MAX_PRELUDE_SIZE
+		// also the size can be zero if after the prelude there is only masked bytes
+		crclen = rz_sign_flirt_crc16_length(mask + RZ_FLIRT_MAX_PRELUDE_SIZE, size - RZ_FLIRT_MAX_PRELUDE_SIZE);
+		crc16 = rz_sign_flirt_crc16(pattern + RZ_FLIRT_MAX_PRELUDE_SIZE, crclen);
+	}
```
Comment (ret2libc) [inline librz/core/cmd/cmd_flirt.c]: having this here seems error prone. I'd expect this kind of low-level details about how flirt works to be hidden from "users" of the flirt API.

File: librz/core/cmd/cmd_flirt.c [hunk 10]
```diff
@@ -41,3 +41,73 @@ RZ_IPI RzCmdStatus rz_flirt_scan_handler(RzCore *core, int argc, const char **ar
...
+
+	for (ut64 i = size - 1; i > 1; --i) {
+		if (mask[i] != 0xFF) {
+			size--;
+			continue;
+		}
+		break;
+	}
+
```
Comment (ret2libc) [inline librz/core/cmd/cmd_flirt.c]: this also seems too low-level. i think these info should come from a proper API.

**Review by ret2libc — COMMENTED** [review]
Some partial comments.

File: librz/core/cmd/cmd_analysis.c (hunk 4 above)
Comment (wargio) [reply librz/core/cmd/cmd_analysis.c]: removed thanks for the info

File: test/db/cmd/cmd_signature (hunk 5 above)
Comment (wargio) [reply test/db/cmd/cmd_signature:205]: it is correct, the match was wrong. i checked manually

File: test/db/cmd/cmd_signature (hunk 6 above)
Comment (wargio) [reply test/db/cmd/cmd_signature:400]: same, the match number is correct, the sig & pat file was made by ida

File: test/db/cmd/cmd_signature (hunk 7 above)
Comment (wargio) [reply test/db/cmd/cmd_signature:408]: because i do not have anymore the original binary, so i just changed it with a new one (yours), also that `data.004011a5` is the main.

File: librz/core/cmd/cmd_flirt.c (hunk 8 above)
Comment (wargio) [reply librz/core/cmd/cmd_flirt.c:65]: no, i checked, when the function is not linear the size is till the jump. this is the correct one.

Comment (wargio) [reply librz/core/cmd/cmd_flirt.c]: The function size is from the entry to the highest address. if you have a jump to an lower address, then the function is terminated to the jump. I have verified this with the output from IDA.

File: test/db/cmd/cmd_signature (hunk 7 above)
Comment (ret2libc) [reply test/db/cmd/cmd_signature:408]: oh :D ok

File: librz/sign/flirt.c [hunk 11]
```diff
@@ -345,6 +347,15 @@ static bool is_pattern_matching(ut32 p_size, const ut8 *pattern, const ut8 *mask
 	return true;
 }
 
+static bool check_crc16(const RzFlirtModule *module, ut8 *b, ut32 b_size) {
+	if (!module->crc_length) {
+		return true;
+	} else if ((b_size - RZ_FLIRT_MAX_PRELUDE_SIZE) < module->crc_length) {
```
Comment (ret2libc) [inline librz/sign/flirt.c:353]: what if b_size is < RZ_FLIRT_MAX_PRELUDE_SIZE? Does it wrap around?

File: librz/core/cmd/cmd_flirt.c [hunk 12]
```diff
@@ -41,3 +41,40 @@ RZ_IPI RzCmdStatus rz_flirt_scan_handler(RzCore *core, int argc, const char **ar
...
+	RzBuffer *buffer = rz_buf_new_empty(0);
+	if (!buffer) {
+		RZ_LOG_ERROR("core: Cannot allocate RzBuffer for writing FLIRT signature\n");
+		rz_sign_flirt_node_free(node);
+		return RZ_CMD_STATUS_ERROR;
+	}
+
+	if (!rz_sign_flirt_write_string_pattern_to_buffer(node, buffer)) {
+		rz_sign_flirt_node_free(node);
```
Comment (ret2libc) [inline librz/core/cmd/cmd_flirt.c:66]: free `buffer` as well.

File: librz/sign/create.c [hunk 13]
```diff
@@ -60,6 +60,17 @@ static RzFlirtFunction *flirt_function_new(const char *name, bool is_local, ut64
 	return function;
 }
 
+static ut32 flirt_crc16_length(RZ_NONNULL const ut8 *mask, size_t size) {
```
Comment (ret2libc) [inline librz/sign/create.c:69]: please add a comment explaining why we compute the flirtcrc length the way we do.

File: librz/sign/create.c [hunk 14]
```diff
@@ -87,14 +98,15 @@ static RzFlirtModule *flirt_module_new(RzAnalysis *analysis, RzAnalysisFunction
 
 	if (b_size > 0 && buffer) {
 		// the crc should be generated only for when the buffer is > RZ_FLIRT_MAX_PRELUDE_SIZE
-		module->crc_length = RZ_MIN(b_size, 0xFF);
+		// also the size can be zero if after the prelude there is only masked bytes
+		module->crc_length = flirt_crc16_length(mask, b_size);
```
Comment (ret2libc) [inline librz/sign/create.c:108]: shouldn't we consider just `mask+RZ_FLIRT_MAX_PRELUDE_SIZE`?

File: librz/sign/flirt.c (hunk 11 above)
Comment (wargio) [reply librz/sign/flirt.c:353]: doesn't matter because it cannot be. it's always at least 32 bytes

File: librz/sign/create.c (hunk 14 above)
Comment (wargio) [reply librz/sign/create.c:108]: no, mask is already shifted by 32 bytes

Comment (wargio) [reply librz/sign/create.c:108]: https://github.com/rizinorg/rizin/blob/8caaad7d17a7448d465ecc8530ec97b2360cf425/librz/sign/create.c#L198-L205

File: librz/sign/flirt.c (hunk 11 above)
Comment (wargio) [reply librz/sign/flirt.c:353]: https://github.com/rizinorg/rizin/blob/8caaad7d17a7448d465ecc8530ec97b2360cf425/librz/sign/flirt.c#L520-L536

**Review by ret2libc — APPROVED** [review]

File: librz/sign/flirt.c (hunk 11 above)
Comment (ret2libc) [reply librz/sign/flirt.c:353]: I'm not sure I follow this. Can you re-explain?
